A87-mtls-spiffe-support.md

```diff
@@ -0,0 +1,188 @@
+the file, or the file contains invalid entries), then there is no valid
+map, which means new connection attempts will fail the TLS handshake.
+For subsequent loads, errors will not affect the existing in-memory
+trust bundle map.  An empty map will be considered a valid (but empty)
+trust bundle map.
+
+Note that the [SPIFFE bundle format] and [Publishing SPIFFE bundle format]
+are partially supported:
+- Only the `keys` and `spiffe_sequence` elements of the JWK set are supported.
+- Only the `kty`, `use`, and `x5c` elements of `keys` are supported.
+- Instead of ignoring individual JWK entries in case of issues, we ignore
+  the whole trust bundle.
+
+### SPIFFE Certificate Verification in TlsCredentials
+
+The xDS mTLS support is built on top of existing mTLS functionality in
+TlsCredentials.  This section describes how the SPIFFE functionality
+will work in TlsCredentials and how the xDS functionality builds on top
+of it.  The details of this are different in each language.
+
+#### C++
+
+In C++, TlsCredentials natively supports a CertificateProvider API with
+essentially the same semantics as in xDS.  We will add the ability for
+the `FileWatcherCertificateProvider` to be instantiated with a SPIFFE
+trust bundle map file instead of a CA certificate file.
+
+TODO: @erm-g to fill in API details.  We'll probably need to introduce
+an options struct for FileWatcherCertificateProvider.
+
+#### Java
+
+In Java, we need new API to work with [SPIFFE] and [SPIFFE bundle
+format].  A new `SpiffeUtil` will be developed:
+
+```java
+class SpiffeUtil {
+   Optional<SpiffeId> extractSpiffeId(X509Certificate[] certChain);
+   SpiffeBundle loadTrustBundleFromFile(String trustBundleFile);
+}
+class SpiffeId {
+   String getTrustDomain();
+   String getPath();
+}
+class SpiffeBundle {
+   Map<String, List<X509Certificates>> getBundleMap();
+}
+```
+
+For the xDS functionality described above, the `XdsX509TrustManager`
+will gain the ability to be instantiated with a SPIFFE trust bundle
+map.  In this case, it will use the SPIFFE trust bundle certificates as
+trusted roots.  If the `XdsX509TrustManager` is instantiated using CA
+certificates (existing functionality), then it'll contimue to use them
+exactly as it does today.  The new APIs will look like this:
+
+```java
+class XdsTrustManagerFactory {
+   XdsTrustManagerFactory(Map<String, List<X509Certificates>>);
+   XdsTrustManagerFactory(X509Certificates[]);
+}
+class XdsX509TrustManager {
+   XdsX509TrustManager(Map<String, XdsX509ExtendedTrustManager>);
+   XdsTrustManagerFactory(XdsX509ExtendedTrustManager);
+}
+```
+
+We'll also adjust existing hierarchies of `Watcher` and
+`CertificateProvider` to support the new SPIFFE trust bundle map.
+
+#### Go
+
+TODO: @erm-g to fill this in.
+
+### Temporary environment variable protection
+
+The xDS functionality will be guarded via the
+`GRPC_EXPERIMENTAL_XDS_MTLS_SPIFFE` environment variable.  The new
```

Comment:
Ugh, `GRPC_EXPERIMENTAL_SPIFFE_TRUST_BUNDLE_MAP` was used in Java. I think that may be because the initial implementation pre-dates this gRFC.

Reply:
I'd prefer to stick with the same env var naming format that we've used for other features.

Can we fix the Java impl?  I don't think we have the interop tests done yet, so this shouldn't break anything to change.

Reply:
Yeah, we can change it for Java.

Reply:
@ejona86 do you want to just make this change or should I handle it?

Reply:
I don't mind making it, but it'll be less likely to be lost if you do it.

Reply:
Our team will take it